File: Data Structures & Algorithms/subsets/submission-3.java
class Solution {
    private List<List<Integer>> res;
    private List<Integer> track;

    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        track = new ArrayList<>();
        backtrack(nums, 0);
        return res;
    }

    private void backtrack(int[] nums, int start) {
        res.add(new ArrayList<>(track));
        for (int i = start; i < nums.length; i++) {
            track.add(nums[i]);
            backtrack(nums, i+1);
            track.remove(track.size() - 1);
        }
    }
}

File: Data Structures & Algorithms/longest-palindromic-substring/submission-3.py
# class Solution:
#     def longestPalindrome(self, s: str) -> str:
#         res = ""
#         n = len(s)
#         for i in range(n):
#             s1 = self.expand(s, i, i)
#             s2 = self.expand(s, i, i+1)
#             res = res if len(res) > len(s1) else s1
#             res = res if len(res) > len(s2) else s2
        
#         return res
    
#     def expand(self, s, l, r):
#         while l >= 0 and r < len(s) and s[l] == s[r]:
#             l -= 1
#             r += 1
        
#         return s[l+1:r]

class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s:
            return ""
        
        start, end = 0, 0
        for i in range(len(s)):
            l1, r1 = self.expand(s, i, i)
            l2, r2 = self.expand(s, i, i+1)
            if r1 - l1 > end - start:
                start, end = l1, r1
            
            if r2 - l2 > end - start:
                start, end = l2, r2
        
        return s[start:end+1]
    
    def expand(self, s, l, r):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        
        return l+1, r-1

